### PR TITLE
JDA Implementation

### PR DESCRIPTION
**Pull request by jjoeldaniel**
### Finished
- [x] Allows users to create/delete triggers and DM when trigger is sent
- [x] Allows for multiple triggers
- [x] Add various bot sub-commands
    - /list, /new, /delete, /reset, /help, /toggle
- [x] Add auto-completion to /delete

### Import Documentation
- [JDA](https://github.com/DV8FromTheWorld/JDA)
- [SLF4J](https://github.com/qos-ch/slf4j)
- [dotenv Java](https://github.com/cdimascio/dotenv-java)
- [fuzzywuzzy](https://github.com/xdrop/fuzzywuzzy)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>acm-csuf</groupId>
```
Comment (EthanThatOneKid) [inline pom.xml]: What do you think about a group ID of `com.acmcsuf.bot-committee`?

File: src/main/java/Register.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    public void onGuildReady(GuildReadyEvent event) {
+
+        // Register commands (#updateCommands will CLEAR all commands, don't do this more than once per startup)
+        event.getGuild().updateCommands().addCommands(guildCommands()).queue();
+    }
+
+    @Override
+    public void onGuildJoin(GuildJoinEvent event) {
+        event.getGuild().updateCommands().addCommands(guildCommands()).queue();
+    }
```
Comment (EthanThatOneKid) [inline src/main/java/Register.java]: Reduce code repetition using a private method, perhaps named `overwriteCommands`.

**Review by EthanThatOneKid — COMMENTED** [review]
I took a look at the PR and left a couple comments :)

File: README.md [hunk 3]
```diff
@@ -1,27 +1,99 @@
...
+## IDE Setup
 
 1. Clone the repository
-1. Create a virtual environment
-1. Install the requirements
-1. Create a `.env` file with the following variables:
+2. Create a `.env` file in the main directory with the following variables:
     - `DISCORD_TOKEN`: The token of the bot
+    - `AUTHORIZED_ROLE_ID`: Array of role IDs that are allowed to use the bot (separate by comma)
+3. Specify JDK version (Tested on JDK 17/19)
```
Comment (EthanThatOneKid) [inline README.md]: Specify JDK 19 since contributors should test on the same version.

File: README.md [hunk 4]
```diff
@@ -1,27 +1,99 @@
...
+
+2. **"I can't run the bot"**
+
+    - Make sure you have the correct token
+    - Make sure you have the correct role IDs
+    - Make sure you have the correct JDK version
+
+3. **"How can I get my bot ID?"**
+
+   - Go to Discord Developer Dashboard -> Your bot -> General Information -> Application ID
```
Comment (EthanThatOneKid) [inline README.md]: Nit: Replace -> with HTML arrow entity → in this file.
```suggestion
   - Go to Discord Developer Dashboard → Your bot → General Information → Application ID
```

File: README.md [hunk 5]
```diff
@@ -1,27 +1,99 @@
...
+## Table of Contents
 
-### Setup
+- [IDE Setup](#ide-setup)
+- [Bot Setup](#bot-setup)
+- [Running](#running)
+- [Commands](#commands)
+- [Documentation](#documentation)
+- [FAQ](#faq)
+- [Contributing](#contributing)
```
Comment (EthanThatOneKid) [inline README.md]: There are 2 target audiences for our README.md: contributors and users. With this in mind, the TOC could be rearranged like so,
```suggestion
- [Usage](#usage)
  - [Commands](#commands)
- [Bot Setup](#bot-setup)
  - [Running](#running)
- [Contributing](#contributing)
  - [IDE Setup](#ide-setup)
  - [Formatting](#formatting)
- [FAQ](#faq)
- [References](#references)
```

- Note: "Documentation" was renamed to "References"
- Note: Add "Formatting" section under "Contributing" about running the project formatter

File: README.md [hunk 6]
```diff
@@ -1,27 +1,99 @@
...
+| `/trigger list`   | Lists all stored triggers  |
+| `/trigger delete` | Deletes specified trigger  |
+| `/trigger toggle` | Toggles trigger feature    |
+
+## Documentation
+
+- <a href='https://github.com/DV8FromTheWorld/JDA'>JDA</a>
+- <a href='https://github.com/qos-ch/slf4j'>SLF4J</a>
+- <a href='https://github.com/cdimascio/dotenv-java'>dotenv Java</a>
+- <a href='https://github.com/xdrop/fuzzywuzzy'>fuzzywuzzy</a>
```
Comment (EthanThatOneKid) [inline README.md]: References used are often helpful for new contributors.
```suggestion
## References

- [JDA](https://github.com/DV8FromTheWorld/JDA)
- [SLF4J](https://github.com/qos-ch/slf4j)
- [dotenv Java](https://github.com/cdimascio/dotenv-java)
- [fuzzywuzzy](https://github.com/xdrop/fuzzywuzzy)
```

File: README.md [hunk 7]
```diff
@@ -1,27 +1,99 @@
...
 
 ## Contributing
 
 1. Create a new branch
-1. Make your changes
-1. Create a pull request
+2. Make your changes
+3. Create a pull request
 
 ---
```
Comment (EthanThatOneKid) [inline README.md:110]: Nit: Feel free to match our `pom.xml` groupId for the salutation at the bottom of this file.

```md
Created with 💖 by **com.acmcsuf.bot-committee**
```

File: src/main/java/Trigger.java [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+    // Discord member ID : Set of trigger phrases
+    HashMap<String, LinkedHashSet<String>> triggerMap = new HashMap<>();
+
+    // Discord member ID : Trigger Activation (true = activated, false = deactivated)
+    HashMap<String, Boolean> triggerToggle = new HashMap<>();
+
+    int min = 0;
+    int max = 5;
+
+    final int MAX_TRIGGERS = 50;
```
Comment (EthanThatOneKid) [inline src/main/java/Trigger.java]: Add more constants `COMMAND_NAME` and `COMMAND_DESCRIPTION` for usage in Register.

**Review by EthanThatOneKid — COMMENTED** [review]
Nice progress (up to 80+ commits in just this PR)! I left a review on this PR as it is nearing completion.

File: src/main/java/Trigger.java (hunk 8 above)
Comment (jjoeldaniel) [reply src/main/java/Trigger.java]: Thoughts?

```Java
public final class Commands
{

    // Command List
    public static final String TRIGGER = "trigger";

    // Subcommands
    public static final String TRIGGER_HELP = "help";
    public static final String TRIGGER_RESET = "reset";
    public static final String TRIGGER_LIST = "list";
    public static final String TRIGGER_NEW = "new";
    public static final String TRIGGER_DELETE = "delete";
    public static final String TRIGGER_TOGGLE = "toggle";

    private Commands()
    {

    }

}
```
Example usage below

```Java
case ( Commands.TRIGGER_HELP ) ->
            {
                EmbedBuilder builder = new EmbedBuilder()
                        .setTitle( "Trigger Commands" )
                        .setColor( Color.green )
                        .addField( "/trigger new", "Add trigger", false )
                        .addField( "/trigger reset", "Removes trigger", false )
                        .addField( "/trigger list", "Lists triggers", false )
                        .addField( "/trigger delete", "Delete trigger", false )
                        .addField( "/trigger toggle", "Toggles trigger feature", false );

                event.replyEmbeds( builder.build() ).setEphemeral( true ).queue();
            }
}
```

Comment (EthanThatOneKid) [reply src/main/java/Trigger.java]: > Thoughts?

I like this :godmode:

File: src/main/java/com/acmcsuf/bot_committee/Register.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                new SubcommandData( "toggle", "Toggles trigger feature" )
+                        .addOption( OptionType.BOOLEAN, "switch", "Toggles feature", true );
+        SubcommandData newTrigger =
+                new SubcommandData( "new", "Add trigger" )
+                        .addOption( OptionType.STRING, "word", "Trigger word", true );
+        SubcommandData delete =
+                new SubcommandData( "delete", "Delete trigger" )
+                        .addOption( OptionType.STRING, "word", "Trigger word", true, true );
+        guildCommandData.add(
+                Commands.slash( "trigger", "Receive a DM when trigger word is mentioned in mutual servers" )
```
Comment (EthanThatOneKid) [inline src/main/java/com/acmcsuf/bot_committee/Register.java]: Use class constant values as slash command name inputs.

Comment (EthanThatOneKid) [reply src/main/java/com/acmcsuf/bot_committee/Register.java]: Fixed in <https://github.com/EthanThatOneKid/crying-counter/pull/2/commits/f3d184d55ad72201e97f6b11be15111a135a5cc6>. Looks nice!

**Review by EthanThatOneKid — APPROVED** [review]
Thanks for all the progress on this big initial PR to the repository; it looks awesome! I’m looking forward to maintaining this program together :)

**Pull request merged**